tee-worker/litentry/core/data-providers/src/moralis.rs

```diff
@@ -0,0 +1,268 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{build_client, DataProviderConfig, Error, HttpError};
+use http::header::CONNECTION;
+use http_req::response::Headers;
+use itc_rest_client::{
+	http_client::{DefaultSend, HttpClient},
+	rest_client::RestClient,
+	RestGet, RestPath,
+};
+use litentry_primitives::Web3Network;
+use log::debug;
+use serde::{Deserialize, Serialize};
+use std::{
+	format, str,
+	string::{String, ToString},
+	thread, time, vec,
+	vec::Vec,
+};
+
+#[derive(Debug)]
+pub struct MoralisRequest {
+	path: String,
+	query: Option<Vec<(String, String)>>,
+}
+
+pub struct MoralisClient {
+	api_retry_delay: u64,
+	api_retry_times: u16,
+	client: RestClient<HttpClient<DefaultSend>>,
+}
+
+impl MoralisClient {
+	pub fn new(data_provider_config: &DataProviderConfig) -> Self {
+		let api_key = data_provider_config.moralis_api_key.clone();
+		let api_retry_delay = data_provider_config.moralis_api_retry_delay;
+		let api_retry_times = data_provider_config.moralis_api_retry_times;
+		let api_url = data_provider_config.moralis_api_url.clone();
+
+		let mut headers = Headers::new();
+		headers.insert(CONNECTION.as_str(), "close");
+		headers.insert("X-API-Key", api_key.as_str());
+		let client = build_client(api_url.as_str(), headers);
+
+		MoralisClient { api_retry_delay, api_retry_times, client }
+	}
+
+	fn retry<A, T>(&mut self, action: A) -> Result<T, Error>
+	where
+		A: Fn(&mut MoralisClient) -> Result<T, HttpError>,
+	{
+		let mut retries = 0;
+		// retry delay 1 second
+		let base_delay = time::Duration::from_millis(self.api_retry_delay);
+		// maximum 5 retry times
+		let maximum_retries = self.api_retry_times;
+
+		loop {
+			if retries > 0 {
+				debug!("Fail to call moralis api, begin retry: {}", retries);
+			}
+
+			if retries > maximum_retries {
+				return Err(Error::RequestError(format!(
+					"Fail to call call moralis api within {} retries",
+					maximum_retries
+				)))
+			}
+
+			match action(self) {
+				Ok(response) => return Ok(response),
+				Err(err) => {
+					let req_err: Error = Error::RequestError(format!("moralis api error: {}", err));
+					match err {
+						HttpError::HttpError(code, _) =>
+							if code == 429 {
+								// Too Many Requests
+								// exponential back off
+								thread::sleep(base_delay * 2u32.pow(retries as u32));
```

Comment:
The exponential backing off here, will make the thread sleep for x seconds, during which the thread is not available for any other computation. Exponential backoff should not be implemented single threaded, Because it essentially is blocking this single thread for x seconds and affecting throughput.

Reply:
In this case should we simply pass this rate limiting error to client side maybe?

Just this is at data-provider layer,  for one VC, the caller could call this method in a for loop too, and if one fail then even we might have some results from other calls, we will have to fail the whole flow.

It's a trade off between throughput and trying server-side's best to make the logic work without requiring client-side retry.

Reply:
See, here's the scenario, in isolated request-vc we have about 4 threads, So if we have 4 threads occupied with an exponential backoff sleep, We potentially have a DOS implemented by ourselves in the code. The ultimate solution is definitely 1 or 2, but until we have that, let's not explicitly sleep(x to the power of n) the entire thread such that no other VCs of other data providers can use the service?

Reply:
Given the thread limitation, I'm not sure if there's a one-size-fits-all solution.
On one hand, if a VC only needs a single data provider request, it's probably best to free up the worker and let the client retry.
On the other hand, if a lot of provider requests are needed (like the holding time assertion, even [after optimizing the logic](github.com/litentry/litentry-parachain/pull/2401), it could be less harmful to retry a couple times rather than forcing the client to restart from scratch.

Reply:
There is never a one size fits solution 😅 . So basically assertion logic that retry are going to take up more of our computation time which increases the response time for other more simpler assertion. Why can't we have temporary storage of already received results and only retry the ones that didn't succeed instead of blocking the thread for **exponential time**.

Reply:
added fail fast boolean option in service layer, and trait in data provider layer now